File: music-api/src/main/java/music/service/DeviceService.java
package music.service;

import music.mapper.DeviceMapper;
import music.model.Device;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class DeviceService {

    private final DeviceMapper deviceMapper;

    @Autowired
    public DeviceService(DeviceMapper deviceMapper) {
        this.deviceMapper = deviceMapper;
    }

    /**
     * If a device with the specified name already exists, return it, otherwise, insert
     * it and then return the newly inserted object.
     */
    public Device getOrInsert(String name) {
        Device device = deviceMapper.getDeviceByName(name);
        if (device == null) {
            deviceMapper.insert(name);
            return deviceMapper.getDeviceByName(name);
        } else {
            return device;
        }
    }
}

File: music-api/src/main/java/music/service/DeviceService.kt
package music.service

import music.mapper.DeviceMapper
import music.model.Device
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class DeviceService @Autowired
constructor(private val deviceMapper: DeviceMapper) {

    /**
     * If a device with the specified name already exists, return it, otherwise, insert
     * it and then return the newly inserted object.
     */
    fun getOrInsert(name: String): Device {
        val device = getDeviceByName(name)
		return if (device == null) {
			deviceMapper.insert(name)
			getDeviceByName(name)!!
		} else {
			device
		}
    }

	fun getDeviceByName(name: String) : Device? = deviceMapper.getDeviceByName(name)
}
